captum/attr/_utils/class_summarizer.py

```diff
@@ -0,0 +1,89 @@
+#!/usr/bin/env python3
+from collections import defaultdict
+from typing import Any, Dict, List, Optional, Tuple, Union
+
+import torch
+from torch import Tensor
+
+from captum.attr._utils.stat import Stat
+from captum.attr._utils.summarizer import Summarizer
+
+
+class ClassSummarizer:
+    """
+    Used to keep track of summaries for associated classes. The
+    classes/labels can be of any type that is supported by `dict`.
+
+    This also keeps track of an aggregate of all class summaries.
+    """
+
+    def __init__(self, stats: List[Stat]):
+        self.stats = stats
+        self.all_summary = Summarizer(stats=stats)
+        self.summaries: Dict[Any, Summarizer] = defaultdict(
+            lambda: Summarizer(stats=stats)
+        )
+
+    def update(
+        self,
+        x: Union[Tensor, Tuple[Tensor, ...]],
+        labels: Optional[Union[Any, List[Any]]] = None,
+    ):
+        """
+        Updates the stats of the summarizer, optionally associated to classes.
+
+        Args:
+            x (Tensor or Tuple[Tensor, ...]):
+                The input tensor to be summarised. The first
+                dimension of this input must be associated to
+                the amount of `labels` specified; unless there
+                is one labels represents one label (in that case
+                the first dimension can optionally be unsqueezed).
+            labels (Any, List[Any], optional):
+                The associated labels for `x`. If Any, we
+                assume x does not represent a batch of inputs.
+
+                If this is None we simply aggregate the total summary.
+        """
+        if labels is not None:
+            if not isinstance(x, tuple):
+                x = (x,)
+
+            should_resqueeze = torch.zeros(len(x))
+            if not isinstance(labels, list):
+                # we need to support the data having
+                # size (1, ...) or (...)
+                labels = [labels]
+                for i in range(len(x)):
+                    should_resqueeze[i] = len(x[i].size()) == 0 or x[i].size(0) != 1
```

Comment:
nit:  I think that `should_resqueeze` variable sounds a bit unnatural. 
Can we possibly check it differently ?Is this check only necessary for Class Summarizer  ? We didn't have it for the original `Summarizer` ?

Reply:
This is done because the `.update` method accepts a single label or multiple labels associated to a batch of tensors. To make the code work for both, for the single label case the input may need to be unsqueezed in order to do the indexing in the for loop; hence I want to potentially re-perform the squeeze to transform it back to what it was.

The alternative is to make `.update` only accept a single label and single tensor, remove the for loop and squeeze/unsqueeze operations.

Reply:
In `Captum` we always assume that the first dimension is the number of examples in the batch. With `torch.numel` you can check if there is only one element in the label tensor.

Otherwise it becomes more tricky and it might make the code more difficult to maintain. For the labels, I'll actually follow the logic we had for target:

https://github.com/pytorch/captum/blob/master/captum/attr/_utils/common.py#L447